src/cursor.c

```diff
@@ -51,26 +47,38 @@ static void Cursor_RemoveFromIdle(Cursor *cur) {
   }
 
   Array_Resize(idle, sizeof(Cursor *) * (n - 1));
-  if (cur->nextTimeoutNs == cur->parent->nextIdleTimeoutNs) {
-    cur->parent->nextIdleTimeoutNs = 0;
+  if (cur->nextTimeoutNs == cl->nextIdleTimeoutNs) {
+    cl->nextIdleTimeoutNs = 0;
   }
   cur->pos = -1;
 }
 
-/* Doesn't lock - simply deallocates and decrements */
+/* Assumed to be called under the cursors global lock or upon server shut down. */
 static void Cursor_FreeInternal(Cursor *cur, khiter_t khi) {
+  CursorList *cl = &g_CursorsList;
   /* Decrement the used count */
-  RS_LOG_ASSERT(khi != kh_end(cur->parent->lookup), "Iterator shouldn't be at end of cursor list");
-  RS_LOG_ASSERT(kh_get(cursors, cur->parent->lookup, cur->id) != kh_end(cur->parent->lookup),
+  RS_LOG_ASSERT(khi != kh_end(cl->lookup), "Iterator shouldn't be at end of cursor list");
+  RS_LOG_ASSERT(kh_get(cursors, cl->lookup, cur->id) != kh_end(cl->lookup),
                                                     "Cursor was not found");
-  kh_del(cursors, cur->parent->lookup, khi);
-  RS_LOG_ASSERT(kh_get(cursors, cur->parent->lookup, cur->id) == kh_end(cur->parent->lookup),
+  kh_del(cursors, cl->lookup, khi);
+  RS_LOG_ASSERT(kh_get(cursors, cl->lookup, cur->id) == kh_end(cl->lookup),
                                                     "Failed to delete cursor");
-  cur->specInfo->used--;
   if (cur->execState) {
     Cursor_FreeExecState(cur->execState);
     cur->execState = NULL;
   }
+  // if There's a spec associated with the cursor
+  if(cur->spec_ref.rm) { 
+    StrongRef spec_ref = WeakRef_Promote(cur->spec_ref);
+    IndexSpec *spec = StrongRef_Get(spec_ref);
+    // the spec may have been dropped, so we need to make sure it is still valid. 
+    if(spec) {
+      spec->activeCursors--;
+      StrongRef_Release(spec_ref);
```

Comment:
The cursor might now keep the index alive for up to `CURSOR_MAX_IDLE` ?

Reply:
No the strong reference is released after each read
We keep the weak ref in the cursor's struct to validate that the spec still exists when we try to call read. If the weak ref promotion fails, the query fails